File: crabzilla-vertx-stack/src/main/java/crabzilla/stack/vertx/verticles/CommandHandlerVerticle.java
package crabzilla.stack.vertx.verticles;

import com.github.benmanes.caffeine.cache.Cache;
import crabzilla.model.AggregateRoot;
import crabzilla.model.Command;
import crabzilla.model.Snapshot;
import crabzilla.model.UnitOfWork;
import crabzilla.model.util.Either;
import crabzilla.stack.EventRepository;
import crabzilla.stack.SnapshotMessage;
import io.vertx.circuitbreaker.CircuitBreaker;
import io.vertx.core.*;
import io.vertx.core.eventbus.Message;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.function.BiFunction;
import java.util.function.Function;

import static crabzilla.model.util.Eithers.getLeft;
import static crabzilla.model.util.Eithers.getRight;
import static crabzilla.stack.util.StringHelper.commandHandlerId;
import static crabzilla.stack.vertx.verticles.CommandExecution.*;
import static java.util.Collections.singletonList;

@Slf4j
public class CommandHandlerVerticle<A extends AggregateRoot> extends AbstractVerticle {

  final Class<A> aggregateRootClass;
  final Function<String, SnapshotMessage<A>> snapshotReaderFn;
  final BiFunction<Command, Snapshot<A>, Either<Exception, Optional<UnitOfWork>>> cmdHandler;
  final Function<Command, List<String>> validatorFn;
  final Cache<String, Snapshot<A>> cache;

  final EventRepository eventRepository;
  final Vertx vertx;
  final CircuitBreaker circuitBreaker;

  @Inject
  public CommandHandlerVerticle(@NonNull final Class<A> aggregateRootClass,
                                @NonNull final Function<String, SnapshotMessage<A>> snapshotReaderFn,
                                @NonNull final BiFunction<Command, Snapshot<A>, Either<Exception, Optional<UnitOfWork>>> cmdHandler,
                                @NonNull final Function<Command, List<String>> validatorFn,
                                @NonNull final EventRepository eventRepository,
                                @NonNull final Cache<String, Snapshot<A>> cache,
                                @NonNull final Vertx vertx,
                                @NonNull @Named("cmd-handler") final CircuitBreaker circuitBreaker) {
    this.aggregateRootClass = aggregateRootClass;
    this.snapshotReaderFn = snapshotReaderFn;
    this.cmdHandler = cmdHandler;
    this.validatorFn = validatorFn;
    this.eventRepository = eventRepository;
    this.cache = cache;
    this.vertx = vertx;
    this.circuitBreaker = circuitBreaker;
  }

  @Override
  public void start() throws Exception {

    vertx.eventBus().consumer(commandHandlerId(aggregateRootClass), msgHandler());

  }

  Handler<Message<Command>> msgHandler() {

    return (Message<Command> msg) -> {

      vertx.executeBlocking((Future<CommandExecution> future) -> {

        val command = msg.body();

        if (command==null) { // TODO optional commandId
          future.complete(VALIDATION_ERROR(UUID.randomUUID(),
                  singletonList("Command cannot be null. Check if JSON payload is valid.")));
          return;
        }

        log.info("received a command {}", command);

        val constraints = validatorFn.apply(command);

        if (!constraints.isEmpty()) {
          future.complete(VALIDATION_ERROR(command.getCommandId(), constraints));
          return;
        }

        circuitBreaker.fallback(throwable -> {

          log.error("Fallback for command " + command.getCommandId(), throwable);
          return FALLBACK(command.getCommandId()); })

        .execute(cmdHandler(command))

        .setHandler(resultHandler(msg));

      }, false, resultHandler(msg));

    };
  }

  Handler<Future<CommandExecution>> cmdHandler(final Command command) {

    return future -> {

      val snapshotDataMsg = snapshotReaderFn.apply(command.getTargetId().getStringValue());

      if (snapshotDataMsg.hasNewSnapshot()) {
        cache.put(command.getCommandId().toString(), snapshotDataMsg.getSnapshot());
      }

      final Either<Exception, Optional<UnitOfWork>> either = cmdHandler.apply(command, snapshotDataMsg.getSnapshot());

      val optException = getLeft(either);

      if (optException.isPresent()) {
        log.error("Business logic error for command " + command.getCommandId(), optException.get());
        future.complete(BUSINESS_ERROR(command.getCommandId()));
        return;
      }

      val optUnitOfWork = getRight(either);

      if (!optUnitOfWork.get().isPresent()) {
        future.complete(UNKNOWN_COMMAND(command.getCommandId()));
        return;
      }

      val uow = optUnitOfWork.get().get();
      val uowSequence = eventRepository.append(uow);
      val cmdHandleResp = SUCCESS(uow, uowSequence);

      future.complete(cmdHandleResp);

    };

  }

  Handler<AsyncResult<CommandExecution>> resultHandler(Message<Command> msg) {

    return (AsyncResult<CommandExecution> resultHandler) -> {

      if (resultHandler.succeeded()) {

        val resp = resultHandler.result();
        log.info("success: {}", resp);
        msg.reply(resp);

      } else {

        log.error("error cause: {}", resultHandler.cause());
        log.error("error message: {}", resultHandler.cause().getMessage());
        resultHandler.cause().printStackTrace();
        // TODO customize conform commandResult
        msg.fail(400, resultHandler.cause().getMessage());
      }

    };
  }
}

File: crabzilla-vertx-core/src/main/java/io/github/crabzilla/vertx/verticles/CommandHandlerVerticle.kt
package io.github.crabzilla.vertx.verticles

import io.github.crabzilla.core.*
import io.github.crabzilla.vertx.*
import io.github.crabzilla.vertx.helpers.EndpointsHelper
import io.vertx.circuitbreaker.CircuitBreaker
import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.eventbus.DeliveryOptions
import io.vertx.core.eventbus.Message
import net.jodah.expiringmap.ExpiringMap
import org.slf4j.LoggerFactory

class CommandHandlerVerticle<A : Entity>(override val name: String,
                                         private val seedValue: A,
                                         private val cmdHandler: (Command, Snapshot<A>) -> CommandResult?,
                                         private val validatorFn: (Command) -> List<String>,
                                         private val snapshotPromoter: SnapshotPromoter<A>,
                                         private val eventJournal: UnitOfWorkRepository,
                                         private val cache: ExpiringMap<String, Snapshot<A>>,
                                         private val circuitBreaker: CircuitBreaker) : CrabzillaVerticle(name, VerticleRole.HANDLER) {

  companion object {
    internal var log = LoggerFactory.getLogger(CommandHandlerVerticle::class.java)
  }

  @Throws(Exception::class)
  override fun start() {

    val consumer = vertx.eventBus().consumer<Command>(EndpointsHelper.cmdHandlerEndpoint(name))

    consumer.handler({ msg ->

      val command = msg.body()

      if (command == null) {

        val cmdExecResult = CommandExecution(commandId = command, result = CommandExecution.RESULT.VALIDATION_ERROR,
          constraints = listOf("Command cannot be null. Check if JSON payload is valid."))
        msg.reply(cmdExecResult)
        return@handler
      }

      log.info("received a command $command")

      val constraints = validatorFn.invoke(command)

      if (!constraints.isEmpty()) {

        val result = CommandExecution(commandId = command.commandId, result = CommandExecution.RESULT.VALIDATION_ERROR,
          constraints = constraints)
        msg.reply(result)
        return@handler
      }

      circuitBreaker.fallback { throwable ->

        log.error("Fallback for command $command.commandId", throwable)
        val cmdExecResult = CommandExecution(commandId = command.commandId, result = CommandExecution.RESULT.FALLBACK)
        msg.reply(cmdExecResult)

      }

        .execute<CommandExecution>(cmdHandler(command))

        .setHandler(resultHandler(msg))

    })

  }

  internal fun cmdHandler(command: Command): (Future<CommandExecution>) -> Unit {

    return { future1 ->

      val targetId = command.targetId.stringValue()

      // get from cache _may_ be blocking if you plug an EntryLoader (from ExpiringMap)
      vertx.executeBlocking<Snapshot<A>>({ fromCacheFuture ->

        log.debug("loading {} from cache", targetId)

        fromCacheFuture.complete(cache[targetId])

      }, false) { fromCacheResult ->

        if (fromCacheResult.failed()) {
          future1.fail(fromCacheResult.cause())
          return@executeBlocking
        }

        val snapshotFromCache = fromCacheResult.result()
        val emptySnapshot = Snapshot(seedValue, 0)
        val cachedSnapshot = snapshotFromCache ?: emptySnapshot

        log.debug("id {} cached lastSnapshotData has version {}. Will check if there any version beyond it",
                targetId, cachedSnapshot)

        val selectAfterVersionFuture = Future.future<SnapshotData>()

        // command handler function _may_ be blocking if your aggregate are using blocking internal services
        eventJournal.selectAfterVersion(targetId, cachedSnapshot.version, selectAfterVersionFuture, name)

        selectAfterVersionFuture.setHandler { fromEventRepoResult ->

          if (fromEventRepoResult.failed()) {
            future1.fail(fromEventRepoResult.cause())
            return@setHandler
          }

          val nonCached = fromEventRepoResult.result()
          val totalOfNonCachedEvents = nonCached.events.size
          log.debug("id {} found {} pending events. Last version is now {}", targetId, totalOfNonCachedEvents,
                  nonCached.version)

          // cmd handler _may_ be blocking. Otherwise, aggregate root would need to use reactive API to call
          // external services
          vertx.executeBlocking<CommandExecution>({ future2 ->

            val resultingSnapshot = if (totalOfNonCachedEvents > 0)
              snapshotPromoter.promote(cachedSnapshot, nonCached.version, nonCached.events)
            else
              cachedSnapshot

            if (totalOfNonCachedEvents > 0) {
              cache[targetId] = resultingSnapshot
            }

            val result = cmdHandler.invoke(command, resultingSnapshot)

            if (result == null) {
              future2.complete(CommandExecution(result = CommandExecution.RESULT.UNKNOWN_COMMAND, commandId = command.commandId))
              return@executeBlocking
            }

            result.inCaseOfSuccess({ uow ->

              val appendFuture = Future.future<Long>()
              eventJournal.append(uow!!, appendFuture, name)

              appendFuture.setHandler { appendAsyncResult ->

                if (appendAsyncResult.failed()) {
                  val error = appendAsyncResult.cause()
                  log.error("appendUnitOfWork for command " + command.commandId, error.message)
                  if (error is DbConcurrencyException) {
                    future2.complete(CommandExecution(result = CommandExecution.RESULT.CONCURRENCY_ERROR, commandId = command.commandId,
                      constraints = listOf(error.message ?: "optimistic locking error")))
                  } else {
                    future2.fail(appendAsyncResult.cause())
                  }
                  return@setHandler
                }

                val finalSnapshot = snapshotPromoter.promote(resultingSnapshot, uow.version, uow.events)
                cache.put(targetId, finalSnapshot)
                val uowSequence = appendAsyncResult.result()
                log.debug("uowSequence: {}", uowSequence)
                future2.complete(CommandExecution(result = CommandExecution.RESULT.SUCCESS, commandId = command.commandId,
                  unitOfWork = uow, uowSequence = uowSequence))
              }
            })

            result.inCaseOfError({ error ->
              log.error("commandExecution", error.message)

                future2.complete(CommandExecution(result = CommandExecution.RESULT.HANDLING_ERROR, commandId = command.commandId,
                  constraints = listOf(error.message ?: "entity handling error")))
            })

          }) { res ->
            if (res.succeeded()) {
              future1.complete(res.result())
            } else {
              log.error("selectAfterVersion ", res.cause())
              future1.fail(res.cause())
            }
          }
        }

      }

    }
  }

  internal fun resultHandler(msg: Message<Command>): (AsyncResult<CommandExecution>) -> Unit {

    return { resultHandler: AsyncResult<CommandExecution> ->
      if (!resultHandler.succeeded()) {
        log.error("resultHandler", resultHandler.cause())
        // TODO customize
        msg.fail(400, resultHandler.cause().message)
      }
      val resp = resultHandler.result()
      val options = DeliveryOptions().setCodecName("CommandExecution")
      msg.reply(resp, options)
      //      msg.reply(resp, options, (Handler<AsyncResult<Message<CommandExecution>>>)
      //              event -> {
      //        if (!event.succeeded()) {
      //          log.error("msg.reply ", event.cause());
      //        }
      //      });
    }
  }

}
